packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +104,224 @@ if (process.env.NODE_ENV === 'development') {
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (visibleCountGuessRange[1] - visibleCountGuessRange[0] === 1 &&
+          visibleCount === visibleCountGuessRange[0])
+      ) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (visibleCountGuessRange[1] === visibleCount && !isOverflowing) {
+        const doubleOfMaxGuess = visibleCountGuessRange[1] * 2;
+
+        setVisibleCountGuessRange([visibleCount, doubleOfMaxGuess]);
+        setVisibleCount(doubleOfMaxGuess);
+
+        return;
+      }
+
+      let newVisibleCountGuessRange = visibleCountGuessRange;
+
+      if (isOverflowing) {
+        // overflowing = we guessed too high. So, new max guess = half the current guess
+        newVisibleCountGuessRange = [visibleCountGuessRange[0], visibleCount];
+      } else {
+        // not overflowing = maybe we guessed too low? So, new min guess = half of current guess
+        newVisibleCountGuessRange = [visibleCount, visibleCountGuessRange[1]];
+      }
+
+      // Next guess is always the middle of the new guess range
+      const newVisibleCount = Math.floor(
+        (newVisibleCountGuessRange[0] + newVisibleCountGuessRange[1]) / 2,
+      );
+
+      // If newVisibleCount is 0 (impossible state as min is 1), end guessing
+      if (newVisibleCount === 0) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      setVisibleCountGuessRange(newVisibleCountGuessRange);
+      setVisibleCount(newVisibleCount);
+    } finally {
+      isGuessing.current = false;
+    }
+  }, [
+    containerRef,
+    isStabilized,
+    itemsCount,
+    orientation,
+    setVisibleCount,
+    visibleCount,
+    visibleCountGuessRange,
+  ]);
+
+  const previousVisibleCount = useLatestRef(visibleCount);
+  const previousVisibleCountGuessRange = useLatestRef(visibleCountGuessRange);
+  const previousContainer = useLatestRef(containerRef.current);
+
+  // Guess each time any of the following changes:
+  // - `visibleCount`
+  // - `visibleCountGuessRange`
+  // - `containerRef`
+  useLayoutEffect(() => {
+    if (isStabilized || isUnitTest) {
+      return;
+    }
+
+    if (
+      visibleCount !== previousVisibleCount.current ||
+      JSON.stringify(visibleCountGuessRange) !==
+        (previousVisibleCountGuessRange.current != null
+          ? JSON.stringify(previousVisibleCountGuessRange.current)
```

Comment:
I have the same question for this check as in [comment](https://github.com/iTwin/iTwinUI/pull/2154/files#r1806739124).

Reply:
Closing this comment since it has been mentioned in [comment](https://github.com/iTwin/iTwinUI/pull/2154#discussion_r1806822405).